### PR TITLE
Builtin: 'LOCATE' dealing with case sensitivity and utf8 characters. 

### PR DESCRIPTION
**Pull request by spongedu**
According to https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_locate, 'LOCATE' is multibyte safe, and is case-sensitive only if at least one argument is a binary string. Current implementation is wrong on both sides.

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -780,25 +781,45 @@ func (b *builtinLocateSig) eval(row []types.Datum) (d types.Datum, err error) {
...
+	}
+	if caseSensitive {
+		if pos < 0 || pos > int64(len(str)-len(subStr)) {
 			d.SetInt64(0)
 			return d, nil
+		} else if len(subStr) == 0 {
+			d.SetInt64(pos + 1)
+			return d, nil
+		}
+		idx = strings.Index(str[pos:], subStr)
```
Comment (hanfei1991) [inline expression/builtin_string.go]: use `idx :=` then you needn't to declare it.

File: expression/builtin_string.go [hunk 2]
```diff
@@ -780,25 +781,45 @@ func (b *builtinLocateSig) eval(row []types.Datum) (d types.Datum, err error) {
...
+		if idx == -1 {
+			ret = 0
```
Comment (hanfei1991) [inline expression/builtin_string.go]: this operation is meaningless, ret's defualt value is 0.

Comment (spongedu) [inline expression/builtin_string.go]: will fix

File: expression/builtin_string.go (hunk 1 above)
Comment (spongedu) [inline expression/builtin_string.go]: ok

**Comment by hanfei1991** [comment]
LGTM

File: expression/builtin_string.go [hunk 3]
```diff
@@ -799,25 +800,40 @@ func (b *builtinLocateSig) eval(row []types.Datum) (d types.Datum, err error) {
...
 		}
 		pos = p - 1
-		if pos < 0 || pos > int64(len(str)) {
+	}
+	var ret int64
+	caseSensitive := false
+	if args[0].Kind() == types.KindBytes || args[1].Kind() == types.KindBytes {
+		caseSensitive = true
+	}
+	if caseSensitive {
```
Comment (tiancaiamao) [inline expression/builtin_string.go]: The if branch and else branch looks the same, can this code be simplified ?

Comment (spongedu) [inline expression/builtin_string.go]: OK, I'll try :)

**Comment by tiancaiamao** [comment]
Oh, this version is not clear than previous one.
I did guess this code works, but not sure:

```
if !caseSentive {
	str = strings.ToLower(str)
	subStr = strings.ToLower(subStr)
}

if pos < 0 || pos > int64(len(str) - len(subStr)) {
	d.SetInt64(0)
	return d, nil
}
if len(subStr) == 0 {
	d.SetInt64(pos+1)
	return d, nil
}
idx := strings.Index(str[pos:], subStr)
if idx != -1 {
	if !caseSentive {
		idx = utf8.RuneCountInString(str[:idx])
	}
	ret = pos + int64(idx) + 1
}

d.SetInt64(ret)
```

Now that it is here, I should not care too much about the details... 
LGTM

**Comment by spongedu** [comment]
@tiancaiamao  if not caseSentive,   we should treat the inputs of 'LOCATE' as []rune, not []byte. So there may be some problem in your code on this. It's a bit difficult to simplify the codes with the type issue. I agree with you that the previous version may be clearer.......

**Pull request merged**